clash-lib/src/Clash/Netlist/Types.hs

```diff
@@ -77,25 +86,145 @@ data TopEntityT = TopEntityT
 
+type FreshCache = HashMap Text (IntMap Word)
+
+-- See: http://vhdl.renerta.com/mobile/source/vhd00037.htm
+--      http://www.verilog.renerta.com/source/vrg00018.htm
+data IdentifierType
+  = Basic
+  -- ^ A basic identifier: does not have to be escaped in order to be a valid
+  -- identifier in HDL.
+  | Extended
+  -- ^ An extended identifier: has to be escaped, wrapped, or otherwise
+  -- postprocessed before writhing it to HDL.
+  deriving (Show, Generic, NFData, Eq)
+
+-- | A collection of unique identifiers. Allows for fast fresh identifier
+-- generation.
+--
+-- __NB__: use the functions in Clash.Netlist.Id. Don't use the constructor directly.
+data IdentifierSet
+  = IdentifierSet {
+      is_allowEscaped :: !Bool
+      -- ^ Allow escaped ids? If set to False, "make" will always behave like
+      -- "makeBasic".
+    , is_hdl :: !HDL
+      -- ^ HDL to generate fresh identifiers for
+    , is_freshCache :: !FreshCache
+      -- ^ Maps an 'i_baseNameCaseFold' to a map mapping the number of
+      -- extensions (in 'i_extensionsRev') to the maximum word at that
+      -- basename/level. For example, if a set would contain the identifiers:
+      --
+      --   [foo, foo_1, foo_2, bar_5, bar_7_8]
+      --
+      -- the map would look like:
+      --
+      --   [(foo, [(0, 0), (1, 2)]), (bar, [(1, 5), (2, 8)])]
+      --
+      -- This mapping makes sure we can quickly generate fresh identifiers. For
+      -- example, generating a new id for "foo_1" would be a matter of looking
+      -- up the base name in this map, concluding that the maximum identifier
+      -- with this basename and this number of extensions is "foo_2",
+      -- subsequently generating "foo_3".
+      --
+      -- Note that an identifier with no extensions is also stored in this map
+      -- for practical purposes, but the maximum ext is invalid.
+    , is_store :: !(HashSet Identifier)
+      -- ^ Identifier store
+    } deriving (Generic, NFData, Show)
+
+-- | HDL identifier. Consists of a base name and a number of extensions. An
+-- identifier with a base name of "foo" and a list of extensions [1, 2] will be
+-- rendered as "foo_1_2".
+--
+-- Note: The Eq instance of "Identifier" is case insensitive! E.g., two
+-- identifiers with base names 'fooBar' and 'FoObAR' are considered the same.
+-- However, identifiers are stored case preserving. This means Clash won't
+-- generate two identifiers with differing case, but it will try to keep
+-- capitalization.
+--
+-- The goal of this data structure is to greatly simplify how Clash deals with
+-- identifiers internally. Any Identifier should be trivially printable to any
+-- HDL.
+--
+-- __NB__: use the functions in Clash.Netlist.Id. Don't use these constructors
+-- directly.
+data Identifier
+  -- | Unparsed identifier. Used for things such as port names, which should
+  -- appear in the HDL exactly as the user specified.
+  = RawIdentifier
+      !Text
+      -- ^ An identifier exactly as given by the user
```

Comment:
The haddock that comes with 8.4 doesn't commenting on bare fields.
You can use this hack if you want: https://github.com/clash-lang/clash-compiler/blob/master/clash-lib/src/Clash/Netlist/BlackBox/Types.hs#L12-L15